### PR TITLE
combine_by_coordinates to handle unnamed data arrays.

### LINKED ISSUE #3248: combine_by_coords fails with DataArrays
#### MCVE Code Sample

```python
da1 = xr.DataArray([1, 2, 3], dims='x', coords={'x': [0, 1, 2]})
da2 = xr.DataArray([3, 4, 5], dims='x', coords={'x': [2, 3, 4]})
xr.combine_by_coords([da1, da2])
```

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-133-d27216ba5688> in <module>
      1 da1 = xr.DataArray([1, 2, 3], dims='x', coords={'x': [0, 1, 2]})
      2 da2 = xr.DataArray([3, 4, 5], dims='x', coords={'x': [2, 3, 4]})
----> 3 xr.combine_by_coords([da1, da2])

~/work/python/xarray/xarray/core/combine.py in combine_by_coords(datasets, compat, data_vars, coords, fill_value, join)
    619         compat=compat,
    620         fill_value=fill_value,
--> 621         join=join,
    622     )
    623 

~/work/python/xarray/xarray/core/merge.py in merge(objects, compat, join, fill_value)
    588             )
    589 
--> 590         obj = obj.to_dataset() if isinstance(obj, DataArray) else obj
    591         dict_like_objects.append(obj)
    592 

~/work/python/xarray/xarray/core/dataarray.py in to_dataset(self, dim, name)
    478             return self._to_dataset_split(dim)
    479         else:
--> 480             return self._to_dataset_whole(name)
    481 
    482     @property

~/work/python/xarray/xarray/core/dataarray.py in _to_dataset_whole(self, name, shallow_copy)
    426         if name is None:
    427             raise ValueError(
--> 428                 "unable to convert unnamed DataArray to a "
    429                 "Dataset without providing an explicit name"
    430             )

ValueError: unable to convert unnamed DataArray to a Dataset without providing an explicit name
```

To get what I want, I need
```xr.combine_by_coords([da1.to_dataset(name='a'), da2.to_dataset(name='a')]).a```

I think the issue is that the code uses `to_dataset` instead of `_to_temp_dataset`

### PR DESCRIPTION
**Pull request by aijams**
- [x] Closes #3248
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
 - [x] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: xarray/tests/test_combine.py
Comment (dcherian) [inline xarray/tests/test_combine.py:689]: I think this should raise an error

File: xarray/tests/test_combine.py
Comment (dcherian) [inline xarray/tests/test_combine.py:675]: We should return a DataArray in this case.

File: xarray/core/combine.py
Comment (dcherian) [inline xarray/core/combine.py:550]: can we also modify `combine_nested` so the two are consistent.

**Review by dcherian — COMMENTED** [review]
Thanks for your patience @aijams 

Sorry for the delay. I've added a few requests.

File: xarray/tests/test_combine.py
Comment (aijams) [reply xarray/tests/test_combine.py:689]: I saw a function (`test_empty_input`) above `test_combine_coords_empty_list` that does exactly the same thing. It looks like this function isn't called anywhere, so I think it should be removed, especially if combining an empty list of datasets should raise an error as you suggest.

File: xarray/tests/test_combine.py
Comment (aijams) [reply xarray/tests/test_combine.py:675]: So the idea is to start with a nested list of DataArrays and end up with a DataArray. This will change the return type from Dataset to (DataArray or Dataset).

Comment (aijams) [reply xarray/tests/test_combine.py:675]: I wonder if `combine_by_coords` should accept a list of both DataArrays and Datasets? In such a case, I believe it should consider the DataArrays as a separate variable to those in the Datasets. The dummy variable would then be merged with the other variables in the resulting Dataset. Otherwise, this function could take either a list of DataArrays or a list of Datasets.

As for `combine_nested`, it doesn't make sense in my mind to have it take both DataArrays and Datasets at the same time, since it aligns its input into a hypercube according to the nested list structure.

I will have `combine_by_coords` take the "either or" approach for now. Let me know if you think it should take both types of input simultaneously.

File: xarray/core/combine.py
Comment (TomNicholas) [inline xarray/core/combine.py:765]: Should we be using `._to_temp_dataset` here? See @dcherian 's [comment here](https://github.com/pydata/xarray/issues/3248#issuecomment-531825386).

File: xarray/tests/test_combine.py
Comment (TomNicholas) [reply xarray/tests/test_combine.py:675]: @aijams the case of a list of both DataArrays and Datasets is the same as@dcherian's case (3) [in this comment](https://github.com/pydata/xarray/issues/3248#issuecomment-526263925). We have to pass the objects down to merge so that it can raise the correct error.

I think it would be nice if we covered the case {DataArrays}->DataArray, but I'm not sure what the easiest way to do that is. If the dataarrays are unnamed, then the combined result will have some sort of default name, which we could check for and then demote to DataArray before returning. But if the DataArrays are named then at the return step I don't think we will know whether the input was originally a DA or DS before it got promoted?

I'm not sure

File: xarray/core/combine.py
Comment (aijams) [reply xarray/core/combine.py:550]: I wrote a test `test_combine_nested_unnamed_data_arrays` that passes a list of unnamed `DataArray`s into `combine_nested` and it produces the expected output. Can you clarify what about `combine_nested` you want to be consistent?

File: xarray/core/combine.py
Comment (max-sixty) [inline xarray/core/combine.py:655]: Is renaming a breaking change?

**Review by max-sixty — COMMENTED** [review]
Thanks @aijams !

@dcherian  knows this so much better than me, so I'll let him take another look.

FYI the whatsnew may need to be shifted to the current version

File: xarray/core/combine.py
Comment (TomNicholas) [reply xarray/core/combine.py:655]: It technically is a breaking change - if someone was previously passing `combine_by_coords(datasets=...)` then this change will break their code. But renaming the argument does make sense with this PR. Not sure whether that means this justifies a deprecation cycle?

Comment (max-sixty) [reply xarray/core/combine.py:655]: I think it probably does. But only a few extra lines and easy to copy from elsewhere.

**Comment by dcherian** [comment]
> @dcherian knows this so much better than me, so I'll let him take another look.

I defer to @TomNicholas =)

File: xarray/core/combine.py
Comment (TomNicholas) [reply xarray/core/combine.py:655]: > I have no idea what justifies a deprecation cycle in your project, or how one is performed.

No worries! (It's very briefly mentioned in our [contributing guide](http://xarray.pydata.org/en/stable/contributing.html#backwards-compatibility), but maybe we should expand that...)

xarray has loads of regular users, and we don't want them to find that downloading a new version of xarray breaks their perfectly good code, even in some minor way. Therefore we normally hold their hand through any changes by warning them in the version before, or possibly by anticipating the way in which we might need to make sure that their old way of using the functions still works temporarily, to give them time to switch to our new way.

In this case the only people who could be affected are people who are currently passing `datasets` as a named argument to `combine_by_coords`, so I think we want to catch that specific possibility and tell them to change the named argument to `data_objects` in future. So perhaps by adding the argument back in with a temporary check like

```python
import warnings

def combine_by_coords(data_objects, ..., datasets=None):

    # TODO remove after version 0.19, see PR4696
    if datasets is not None:
        warnings.warn("The datasets argument has been renamed to `data_objects`. In future passing a value for datasets will raise an error.")
        data_objects = datasets
```

Does that make sense?

Comment (aijams) [reply xarray/core/combine.py:655]: That makes sense. I'm somewhat concerned about users who ignore the warning (many software projects I've seen generate a lot of warnings), however I don't think there's much that can be done since the method signature will have to change anyway.

File: doc/whats-new.rst
Comment (TomNicholas) [inline doc/whats-new.rst:413]: Are the rest of these changes supposed to be included in this PR?

File: xarray/core/combine.py
Comment (TomNicholas) [inline xarray/core/combine.py:561]: This is a nice way to refactor, that makes it easier to reason about the code.

File: xarray/core/combine.py
Comment (TomNicholas) [inline xarray/core/combine.py:584]: Can this Error ever be reached? Wouldn't the 

```python
if not data_objects:
        return Dataset()
```
stop this from being triggered?

File: xarray/tests/test_combine.py
Comment (TomNicholas) [inline xarray/tests/test_combine.py:725]: Do we need a similar test for `combine_nested`? Even if it passes through fine?

**Review by TomNicholas — CHANGES_REQUESTED** [review]
This is a really nice PR @aijams ! Sorry for taking a while to look at it. I just have a couple of comments.

File: xarray/core/combine.py
Comment (TomNicholas) [reply xarray/core/combine.py:620]: I think I know what @dcherian meant - at first glance it looks like the `_combine_single_variable_hypercube` refactoring is a change that would also simplify the code in `combine_nested`. But looking more closely I don't think it actually makes sense to do that, does it? It seems about as neat as it can be as is.

File: xarray/core/combine.py
Comment (aijams) [reply xarray/core/combine.py:584]: I intended this check to be part of the internal contract for `_combine_single_variable_hypercube` since it doesn't make sense to build an empty hypercube with this method. There would be no way to determine the variable involved, so a default sentinel name would have to be introduced.

File: xarray/tests/test_combine.py
Comment (aijams) [reply xarray/tests/test_combine.py:725]: I could add a similar test for `combine_nested`. Right now `combine_nested` will raise `AttributeError` when run with the test above since the DataArray doesn't have a name.

Comment (TomNicholas) [reply xarray/tests/test_combine.py:725]: We should crystallise what we expect the behaviour to be in the form of a test. If the `AttributeError` currently returned is informative, then that's fine, if it's not then we should consider changing something to raise a better message.

File: doc/whats-new.rst
Comment (aijams) [reply doc/whats-new.rst:413]: No this was an accidental duplication. Sorry about that

File: xarray/tests/test_combine.py
Comment (aijams) [reply xarray/tests/test_combine.py:725]: I decided that the `AttributeError` currently returned is rather confusing. I would rather declare up front in the combine* functions that such a mixed input (both unnamed `DataArray`s and `Dataset`s) is not allowed. This way a user of these functions can pretty quickly see that they are passing in an invalid set of inputs. Both functions now return `ValueError` along with an informative message.

File: xarray/core/combine.py
Comment (TomNicholas) [inline xarray/core/combine.py:636]:
```suggestion
    data_objects,
```

(It's [considered bad practice](https://nikos7am.com/posts/mutable-default-arguments/) to have mutable default arguments to functions in python.)

File: xarray/tests/test_combine.py
Comment (TomNicholas) [inline xarray/tests/test_combine.py:683]:
```suggestion
            ValueError, match=r"Can't combine datasets with unnamed dataarrays."
```

Tiny clarification that this means datasets with other xarray.datarrays, not something about the numpy arrays inside the xarray.dataset objects.

**Review by TomNicholas — APPROVED** [review]
I think that (aside from two miniscule suggestions) this is ready to merge!

File: xarray/core/combine.py
Comment (aijams) [reply xarray/core/combine.py:636]: I put this in because if someone calls this method with datasets as a named parameter the data_objects argument would be unspecified and their code would break with an unspecified argument error. This is part of the deprecation warning below.

Comment (TomNicholas) [reply xarray/core/combine.py:636]: You make a good point, but that means the default argument should be `None`, not an empty list, as `None` is immutable.

File: xarray/tests/test_combine.py
Comment (aijams) [reply xarray/tests/test_combine.py:683]: Good clarification.

**Comment by aijams** [comment]
Sorry about the closing/reopening. I thought I was supposed to do something to get the code merged in to the official master. I accidentally pressed the wrong button thinking it did something else.

**Comment by aijams** [comment]
It looks like this PR has been neglected for a while. I haven't heard from a maintainer whether they will be merging in this PR any time soon.

**Comment by TomNicholas** [comment]
Sorry @aijams ! I was following our broad rule of "two maintainers approve before merging", but this has sat here for absolutely ages and had feedback from others previously, so maybe I will just merge it now.

We really need a better system for catching PRs that just sit around :/

**Comment by max-sixty** [comment]
Sorry @aijams ! Thank you for the reminder and forgive the delay. We appreciate the contribution (more than our responsiveness suggests!)

File: doc/whats-new.rst
Comment (dcherian) [inline doc/whats-new.rst:121]: this is in the wrong place! cc @TomNicholas

Comment (TomNicholas) [reply doc/whats-new.rst:121]: Oops - should now be fixed in `main`

**Pull request merged**
